src/storage/pooled_storage_manager.h

```diff
@@ -88,7 +92,7 @@ class GPUPooledStorageManager final : public StorageManager {
 };  // class GPUPooledStorageManager
 
 void GPUPooledStorageManager::Alloc(Storage::Handle* handle) {
-  std::lock_guard<std::mutex> lock(mutex_);
+  std::lock_guard<std::mutex> lock(Storage::Get()->GetMutex(Context::kGPU));
```

Comment:
So now memory allocation on all gpus share the same mutex? This could slow down memory allocation. Especially when using Gluon.

Reply:
Could you suggest a Gluon benchmark to check the performance impact? If the impact is too big we can move to cooperative launch in NCCL2 (that would not need a shared mutex anymore), but this would mean it will be compatible only with NCCL 2 and CUDA 9. Also, currently cooperative launch is slower than parallel.

Reply:
I may be looking for places to insert many-read/single-write shared mutexes in places to help performance. Do you think that this would be a good candidate for this, or is there a reason that there needs to be a global lock for this operation?

Reply:
There is a reason it needs to be global lock - NCCL needs to finish scheduling all kernels before anybody can start allocation/deallocation of gpu memory, otherwise deadlock will happen.

Reply:
For multiple GPUs, would this just be applicable to the two GPU's involved in the transfer?  Or just one GPU? Or all?

Reply:
NCCL is collective communication library so all gpus are involved at the same time.

Reply:
Thank you for the clarification